chatsky/core/context.py

```diff
@@ -146,108 +146,241 @@ class Context(BaseModel):
     def last_label(self) -> AbsoluteNodeLabel:
         """
-        Return the last :py:class:`~.AbsoluteNodeLabel` of
-        the :py:class:`~.Context`.
-
-        :raises ContextError: If :py:attr:`labels` is empty.
+        Receive last turn label.
+        Throw an error if no labels are present or the last label is absent.
+        :return: The last turn label.
         """
+
         if len(self.labels) == 0:
-            raise ContextError("Labels are empty. Use `Context.init` to initialize context with labels.")
-        last_index = get_last_index(self.labels)
-        return self.labels[last_index]
+            raise ContextError("Labels are empty.")
+        return self.labels._items[self.labels.keys()[-1]]
 
     @property
-    def last_response(self) -> Optional[Message]:
+    def last_response(self) -> Message:
         """
-        Return the last response of the current :py:class:`~.Context`.
-        Return ``None`` if no responses have been added yet.
+        Receive last turn response.
+        Throw an error if no responses are present or the last response is absent.
+        :return: The last turn response.
         """
+
         if len(self.responses) == 0:
-            return None
-        last_index = get_last_index(self.responses)
-        response = self.responses[last_index]
-        return response
+            raise ContextError("Responses are empty.")
+        return self.responses._items[self.responses.keys()[-1]]
 
     @property
     def last_request(self) -> Message:
         """
-        Return the last request of the current :py:class:`~.Context`.
-
-        :raises ContextError: If :py:attr:`responses` is empty.
+        Receive last turn request.
+        Throw an error if no requests are present or the last request is absent.
+        :return: The last turn request.
         """
+
         if len(self.requests) == 0:
-            raise ContextError("No requests have been added.")
-        last_index = get_last_index(self.requests)
-        return self.requests[last_index]
+            raise ContextError("Requests are empty.")
+        return self.requests._items[self.requests.keys()[-1]]
 
     @property
     def pipeline(self) -> Pipeline:
-        """Return :py:attr:`.FrameworkData.pipeline`."""
+        """
+        Return :py:attr:`.FrameworkData.pipeline`.
+        """
+
         pipeline = self.framework_data.pipeline
         if pipeline is None:
             raise ContextError("Pipeline is not set.")
         return pipeline
 
     @property
     def current_node(self) -> Node:
-        """Return :py:attr:`.FrameworkData.current_node`."""
+        """
+        Return :py:attr:`.FrameworkData.current_node`.
+        """
+
         node = self.framework_data.current_node
         if node is None:
             raise ContextError("Current node is not set.")
         return node
+
+    async def turns(self, key: Union[int, slice]) -> Iterable[Tuple[AbsoluteNodeLabel, Message, Message]]:
```

Comment:
Make turns a property that returns an instance of a `Turns` class with `async __getitem__`.
So that the usage of `turns` is `await ctx.turns[1:3]` instead of `await ctx.turns(slice(1, 3))`.

Also, make it so that `turns[1]` is a single tuple while `turns[1:2]` is an iterable of tuples.